File: backend/api/routes/consult.py
"""C端咨询模块 SSE 路由 — /api/v1/consult/messages。

双层检索 (SQL + RAG) + 后置校验 + 失败重生成。
会话隔离：仅接受 sess_consult_ 前缀的 session_id。
"""
import asyncio
import json
import logging
import uuid

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI
from sqlalchemy import select

from config import settings
from models import async_session
from models.college import College
from schemas.consult import ConsultMessageRequest
from services.consult_retrieval_service import (
    query_admission_data,
    build_rag_query,
    render_admission_table,
)
from services.consult_service import (
    CONSULT_SESSION_PREFIX,
    get_session,
    get_chat_history,
    save_message,
    build_profile_summary,
)
from services.consult_validator import validate_response
from services.consult_intent_service import extract_intent
from services.prompt_service import load_prompt
from services.persona_service import build_persona_greeting, apply_persona_style
from tenants.service import resolve_tenant
from core.event_writer import write_event
from utils.jwt import decode_token

router = APIRouter(prefix="/api/v1", tags=["consult"])

_logger = logging.getLogger(__name__)


def _sse(event_type: str, data: dict) -> str:
    """格式化 SSE data 行。"""
    payload = {"type": event_type, **data}
    return f"data: {json.dumps(payload, ensure_ascii=False)}\n\n"


@router.post("/consult/messages")
async def send_consult_message(body: ConsultMessageRequest, request: Request):
    """咨询 SSE 流式端点。

    SSE 事件序列：
      thinking → intent → admission_data → search_start → source* → search_end
      → token* → validation → [regeneration → token*] → done
    """
    # 1. 会话校验：必须是咨询会话（sess_consult_ 前缀）
    if not body.session_id.startswith(CONSULT_SESSION_PREFIX):
        return StreamingResponse(
            iter([_sse("error", {"code": "INVALID_SESSION", "message": "非咨询会话"})]),
            media_type="text/event-stream",
        )

    session = await get_session(body.session_id)
    if not session:
        return StreamingResponse(
            iter([_sse("error", {"code": "SESSION_NOT_FOUND", "message": "会话不存在或已过期"})]),
            media_type="text/event-stream",
        )

    tenant = await resolve_tenant(body.tenant_slug)
    tenant_id = tenant.id if tenant else None

    # 解析 JWT（可选）
    user_id = None
    auth_header = request.headers.get("Authorization", "")
    if auth_header.startswith("Bearer "):
        try:
            payload = decode_token(auth_header[7:])
            if payload:
                user_id = uuid.UUID(payload["user_id"])
        except Exception:
            pass

    user_content = body.message
    await save_message(body.session_id, "user", user_content)

    if tenant_id:
        try:
            await write_event(
                tenant_id, "consult_message_sent",
                session_id=session.id,
                payload={"message_length": len(user_content)},
            )
        except Exception as e:
            _logger.warning(f"Event consult_message_sent failed: {e}")

    # 预加载 tenant college_id
    tenant_college_id = None
    if tenant:
        try:
            async with async_session() as db:
                result = await db.execute(
                    select(College).where(College.name == tenant.config.get("brand", {}).get("name", "华南师范大学"))
                )
                college = result.scalar_one_or_none()
                if college:
                    tenant_college_id = college.id
        except Exception as e:
            _logger.warning(f"Failed to resolve college_id for tenant {body.tenant_slug}: {e}")

    async def event_stream():
        # ── Phase 1: 意图提取（三阶段管道：规则 + LLM + 融合）──
        yield _sse("thinking", {"status": "正在理解你的问题..."})

        # 获取对话历史与画像，供意图提取消解指代
        history_msgs = await get_chat_history(body.session_id, limit=8)
        slots = build_profile_summary(session) or {}

        intent_obj = await extract_intent(
            user_content=user_content,
            tenant_slug=body.tenant_slug,
            history=history_msgs,
            slots=slots,
        )
        intent = intent_obj.to_dict()

        yield _sse("intent", {"intent": intent})

        # ── Phase 2a: SQL 精确查询 admission_data ──
        admission_rows = []
        if intent.get("need_admission_data") and intent.get("majors") and tenant_college_id:
            try:
                province = intent.get("province") or session.province or "广东"
                admission_rows = await query_admission_data(
                    majors=intent["majors"],
                    province=province,
                    year=intent.get("year"),
                    tenant_college_id=tenant_college_id,
                )
            except Exception as e:
                _logger.warning(f"query_admission_data failed: {e}")

        yield _sse("admission_data", {"count": len(admission_rows), "rows": admission_rows})

        # ── Phase 2b: RAG 向量检索 ──
        sources = []
        rag_query = build_rag_query(intent, user_content)
        if rag_query:
            yield _sse("search_start", {})
            try:
                from knowledge_base.chroma_client import search_similar
                loop = asyncio.get_running_loop()
                results = await loop.run_in_executor(
                    None, search_similar, rag_query, 5, body.tenant_slug
                )
                for i, r in enumerate(results[:5]):
                    src = {
                        "text": r["document"][:200],
                        "source_title": r.get("metadata", {}).get("source_title", ""),
                        "source_url": r.get("metadata", {}).get("source_url", ""),
                        "score": round(1 - r.get("distance", 0), 4),
                    }
                    sources.append(src)
                    yield _sse("source", {"index": i, "total": len(results[:5]), "item": src})
            except Exception as e:
                _logger.warning(f"RAG search failed: {e}")
            yield _sse("search_end", {"count": len(sources)})

        # ── Phase 3: 构建 system prompt + LLM 流式生成 ──
        system_template = await load_prompt("consult_system", body.tenant_slug)

        # slots_summary
        existing_profile = build_profile_summary(session) or {}
        province = existing_profile.get("province") or "未知"
        subjects = existing_profile.get("subjects") or "未知"
        score = existing_profile.get("score") or "未知"
        rank = existing_profile.get("rank") or "未知"
        slots_text = f"省份: {province}, 选科: {subjects}, 分数: {score}, 位次: {rank}"

        admission_table = render_admission_table(admission_rows)

        knowledge_context = ""
        if sources:
            lines = ["## 知识库检索结果 (仅供参考)"]
            for i, s in enumerate(sources, 1):
                lines.append(f"{i}. {s['text']}")
            knowledge_context = "\n".join(lines)

        try:
            system_content = system_template.format(
                slots_summary=slots_text,
                admission_table=admission_table,
                knowledge_context=knowledge_context,
            )
        except KeyError as e:
            _logger.warning(f"System prompt template missing placeholder: {e}")
            system_content = system_template

        # 注入租户 AI 形象（assistant_name + greeting + style）
        # 咨询模块与推荐模块共用同一份 persona 配置，确保学生侧体验一致
        if tenant and tenant.config:
            _brand = tenant.config.get("brand", {})
            _uni_short = _brand.get("short_name") or _brand.get("name", "")
            _persona = tenant.config.get("ai_persona", {})
            if _uni_short:
                system_content = build_persona_greeting(_persona, _uni_short) + "\n\n" + system_content
            system_content = apply_persona_style(system_content, _persona)

        # 历史消息
        history_msgs = await get_chat_history(body.session_id, limit=10)
        history = []
        for m in history_msgs:
            if m["role"] == "user":
                history.append(HumanMessage(content=m["content"]))
            else:
                history.append(AIMessage(content=m["content"]))
        # 移除最后一条 user 消息（避免重复）
        if history and isinstance(history[-1], HumanMessage):
            history.pop()

        msgs = [SystemMessage(content=system_content)] + history + [HumanMessage(content=user_content)]

        # LLM 流式生成（temperature=0.3 更严谨）
        llm_gen = ChatOpenAI(
            model=settings.deepseek_model,
            api_key=settings.deepseek_api_key,
            base_url=settings.deepseek_base_url,
            temperature=0.3,
        )

        full_content = ""
        try:
            async for chunk in llm_gen.astream(msgs):
                token = chunk.content if hasattr(chunk, "content") else str(chunk)
                full_content += token
                yield _sse("token", {"text": token})
        except Exception as exc:
            _logger.error(f"LLM stream failed: {exc}")
            yield _sse("error", {"code": "LLM_FAILED", "message": "AI 服务暂时不可用"})
            return

        # ── Phase 4: 后置校验 + 重生成 ──
        issues = []
        regenerated = False
        try:
            issues = validate_response(full_content, admission_rows)
        except Exception as e:
            _logger.warning(f"Validation failed: {e}")

        if issues and admission_rows:
            yield _sse("validation", {
                "passed": False,
                "issues_count": len(issues),
                "issues": [
                    {
                        "major": i.major_in_reply,
                        "metric": i.metric,
                        "value": i.value_in_reply,
                        "issue_type": i.issue_type,
                    }
                    for i in issues
                ],
                "regenerated": False,
            })

            # 重生成（最多 1 次）
            max_attempts = settings.consult_max_regeneration_attempts
            for attempt in range(max_attempts):
                try:
                    degraded_prompt = await load_prompt("consult_degraded", body.tenant_slug)
                    try:
                        degraded_content = degraded_prompt.format(
                            admission_table=admission_table,
                            user_content=user_content,
                        )
                    except KeyError:
                        degraded_content = degraded_prompt

                    regen_msgs = [
                        SystemMessage(content=degraded_content),
                        HumanMessage(content=user_content),
                    ]

                    full_content = ""
                    yield _sse("regeneration", {"attempt": attempt + 1})
                    async for chunk in llm_gen.astream(regen_msgs):
                        token = chunk.content if hasattr(chunk, "content") else str(chunk)
                        full_content += token
                        yield _sse("token", {"text": token})

                    regenerated = True
                    # 重新校验
                    new_issues = validate_response(full_content, admission_rows)
                    if not new_issues:
                        issues = []
                        break
                    issues = new_issues
                except Exception as e:
                    _logger.warning(f"Regeneration attempt {attempt + 1} failed: {e}")
                    break

            yield _sse("validation", {
                "passed": len(issues) == 0,
                "issues_count": len(issues),
                "issues": [
                    {
                        "major": i.major_in_reply,
                        "metric": i.metric,
                        "value": i.value_in_reply,
                        "issue_type": i.issue_type,
                    }
                    for i in issues
                ],
                "regenerated": regenerated,
            })
        else:
            yield _sse("validation", {
                "passed": True,
                "issues_count": 0,
                "issues": [],
                "regenerated": False,
            })

        # ── Phase 5: 保存 assistant 消息 + done ──
        assistant_msg = await save_message(body.session_id, "assistant", full_content)

        if tenant_id:
            try:
                await write_event(
                    tenant_id, "consult_response_completed",
                    session_id=session.id,
                    payload={
                        "response_length": len(full_content),
                        "validation_passed": len(issues) == 0,
                        "regenerated": regenerated,
                    },
                )
            except Exception as e:
                _logger.warning(f"Event consult_response_completed failed: {e}")

        yield _sse("done", {
            "session_id": body.session_id,
            "assistant_message": assistant_msg,
            "validation_passed": len(issues) == 0,
            "regenerated": regenerated,
        })

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: mini-app/src/stores/consult.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { api, getToken } from "@/utils/api";
import { TENANT_SLUG } from "@/utils/config";
import {
  getConsultSessionId,
  saveConsultSessionId,
  clearConsultSessionId,
} from "@/utils/consultSession";

const API_BASE =
  process.env.NODE_ENV === "development"
    ? "/api/v1"
    : (import.meta.env.VITE_API_BASE_URL as string) || "/api/v1";

export interface ConsultSource {
  title: string;
  url: string;
  text?: string;
}

export interface ConsultMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  kind?: "answer" | "validation_warning";
  regenerated?: boolean;
  validationPassed?: boolean;
  sources?: ConsultSource[];
}

// 默认开场白（结合知识库项目信息：KB001/003/005-007/008-012/017-019/021-023/030）
const DEFAULT_GREETING =
  "你好，我是华南师范大学国际商学院出国留学项目咨询助手。\n\n本系统仅提供国际留学项目相关咨询，涵盖以下内容：\n• 2+2 出国留学培训项目（商科/新媒体方向）\n• 3+1 SQA-AD 项目（市场营销/人力资源管理/商务会计）\n• 招生报名、入学考试、学费奖学金、学历认证\n• 合作院校、专业设置、升学就业\n\n你可以直接提问，我会基于学校官方信息为你解答。";

export const useConsultStore = defineStore("consult", () => {
  const sessionId = ref<string | null>(getConsultSessionId());
  const messages = ref<ConsultMessage[]>([]);
  const isLoading = ref(false);
  const validationWarning = ref<string | null>(null);
  const greeting = ref<string | null>(null);
  const assistantName = ref<string | null>(null);

  /** 初始化或恢复咨询会话 */
  async function enterSession(): Promise<boolean> {
    const token = getToken();
    if (!token) return false;

    // 每次进入重新读取存储，避免 logout 清存储后内存 ref 仍持有旧 session_id
    const storedId = getConsultSessionId();
    sessionId.value = storedId;

    try {
      const result = await api.post<{
        session_id: string;
        chat_history?: Array<{ message_id: string; role: string; content: string }>;
        greeting?: string | null;
        assistant_name?: string | null;
      }>("/miniapp/enter", {
        session_id: storedId || null,
        tenant_slug: TENANT_SLUG,
        module_type: "consult",
      });

      if (result.data?.session_id) {
        sessionId.value = result.data.session_id;
        saveConsultSessionId(result.data.session_id);
        if (result.data.greeting) greeting.value = result.data.greeting;
        if (result.data.assistant_name) assistantName.value = result.data.assistant_name;
        if (result.data.chat_history?.length) {
          messages.value = result.data.chat_history.map((m) => ({
            id: m.message_id,
            role: m.role as "user" | "assistant",
            content: m.content,
          }));
        } else {
          // 新会话：插入开场白（后端 greeting 优先，否则使用默认开场白）
          messages.value = [{
            id: `greeting_${Date.now()}`,
            role: "assistant",
            content: greeting.value || DEFAULT_GREETING,
            kind: "answer",
          }];
        }
        return true;
      }
      return false;
    } catch (e) {
      console.error("Consult enter failed:", e);
      return false;
    }
  }

  /** SSE 流式发送消息 */
  async function sendMessage(content: string): Promise<void> {
    if (!sessionId.value || !content.trim()) return;

    const userMsg: ConsultMessage = {
      id: `u_${Date.now()}`,
      role: "user",
      content,
    };
    messages.value.push(userMsg);

    const aiMsg: ConsultMessage = {
      id: `a_${Date.now()}`,
      role: "assistant",
      content: "",
      kind: "answer",
      sources: [],
    };
    messages.value.push(aiMsg);

    isLoading.value = true;
    validationWarning.value = null;

    try {
      const resp = await fetch(`${API_BASE}/consult/messages`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${getToken()}`,
          "X-Tenant": TENANT_SLUG,
        },
        body: JSON.stringify({
          session_id: sessionId.value,
          tenant_slug: TENANT_SLUG,
          message: content,
        }),
      });

      if (!resp.ok || !resp.body) {
        throw new Error(`HTTP ${resp.status}`);
      }

      const reader = resp.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });

        // 后端 SSE 格式：data: {...}\n\n（无 event: 行，type 嵌在 JSON）
        const events = buffer.split("\n\n");
        buffer = events.pop() || "";

        for (const evt of events) {
          const line = evt.trim();
          if (!line.startsWith("data: ")) continue;
          const dataStr = line.slice(6);
          try {
            const data = JSON.parse(dataStr);
            handleSSEEvent(data.type, data, aiMsg);
          } catch (e) {
            console.warn("SSE parse failed:", e, dataStr);
          }
        }
      }
    } catch (e: any) {
      aiMsg.content = "抱歉，AI 服务暂时不可用，请稍后重试。";
      console.error("Consult SSE failed:", e);
    } finally {
      isLoading.value = false;
    }
  }

  function handleSSEEvent(type: string, data: any, aiMsg: ConsultMessage) {
    switch (type) {
      case "thinking":
        // UI 可显示"正在理解你的问题..."
        break;
      case "intent":
        // 意图抽取结果（可选展示）
        break;
      case "admission_data":
        // SQL 检索结果计数
        break;
      case "search_start":
        // RAG 检索中
        break;
      case "source":
        if (aiMsg.sources && data.item) {
          aiMsg.sources.push({
            title: data.item.source_title || "",
            url: data.item.source_url || "",
            text: data.item.text || "",
          });
        }
        break;
      case "search_end":
        break;
      case "token":
        // 增量累积 token（后端字段是 text）
        aiMsg.content += data.text || "";
        break;
      case "validation":
        if (!data.passed) {
          aiMsg.kind = "validation_warning";
          aiMsg.validationPassed = false;
          const issueCount = data.issues_count || 0;
          validationWarning.value =
            issueCount > 0
              ? `本次回答中有 ${issueCount} 项数据未通过校验，正在重新生成...`
              : "本次回答中的部分数据未经系统校验通过，请核对官方来源";
        } else {
          aiMsg.validationPassed = true;
          if (data.regenerated) {
            aiMsg.regenerated = true;
            validationWarning.value = null;
          }
        }
        break;
      case "regeneration":
        // 清空当前 content 准备接收重生成
        aiMsg.content = "";
        break;
      case "done":
        if (data.assistant_message?.message_id) {
          aiMsg.id = data.assistant_message.message_id;
        }
        if (data.validation_passed === false) {
          aiMsg.kind = "validation_warning";
          aiMsg.validationPassed = false;
          validationWarning.value =
            "本次回答仍存在数据校验未通过项，请参考官方招生章程核对";
        }
        break;
      case "error":
        aiMsg.content = data.message || "AI 服务暂时不可用";
        break;
    }
  }

  function clearSession() {
    clearConsultSessionId();
    sessionId.value = null;
    messages.value = [];
    validationWarning.value = null;
    greeting.value = null;
    assistantName.value = null;
  }

  return {
    sessionId,
    messages,
    isLoading,
    validationWarning,
    greeting,
    assistantName,
    enterSession,
    sendMessage,
    clearSession,
  };
});
